cmd/cmd_test.go

```diff
@@ -42,13 +74,315 @@ func (s *cmdSuite) TestDistroSupportsReExec(c *C) {
+}
+
+func (s *cmdSuite) TestCoreSupportsReExecBadInfo(c *C) {
+	// can't read snapd/info if it's a directory
+	d := c.MkDir()
+	p := d + "/usr/lib/snapd/info"
+	c.Assert(os.MkdirAll(p, 0755), IsNil)
+
+	c.Check(cmd.CoreSupportsReExec(d), Equals, false)
+}
+
+func (s *cmdSuite) TestCoreSupportsReExecBadInfoContent(c *C) {
+	// can't understand snapd/info if all it holds are potatoes
+	d := c.MkDir()
+	p := d + "/usr/lib/snapd"
+	c.Assert(os.MkdirAll(p, 0755), IsNil)
+	c.Assert(ioutil.WriteFile(p+"/info", []byte("potatoes"), 0644), IsNil)
+
+	c.Check(cmd.CoreSupportsReExec(d), Equals, false)
+}
+
+func (s *cmdSuite) TestCoreSupportsReExecBadVersion(c *C) {
+	// can't understand snapd/info if all its version is gibberish
+	d := c.MkDir()
+	p := d + "/usr/lib/snapd"
+	c.Assert(os.MkdirAll(p, 0755), IsNil)
+	c.Assert(ioutil.WriteFile(p+"/info", []byte("VERSION=0:"), 0644), IsNil)
+
+	c.Check(cmd.CoreSupportsReExec(d), Equals, false)
+}
+
+func (s *cmdSuite) TestCoreSupportsReExecOldVersion(c *C) {
+	// can't re-exec if core version is too old
+	defer cmd.MockVersion("2")()
+	d := c.MkDir()
+	p := d + "/usr/lib/snapd"
+	c.Assert(os.MkdirAll(p, 0755), IsNil)
+	c.Assert(ioutil.WriteFile(p+"/info", []byte("VERSION=0"), 0644), IsNil)
+
+	c.Check(cmd.CoreSupportsReExec(d), Equals, false)
+}
+
+func (s *cmdSuite) TestCoreSupportsReExec(c *C) {
+	defer cmd.MockVersion("2")()
+	d := c.MkDir()
+	p := d + "/usr/lib/snapd"
+	c.Assert(os.MkdirAll(p, 0755), IsNil)
+	c.Assert(ioutil.WriteFile(p+"/info", []byte("VERSION=9999"), 0644), IsNil)
+
+	c.Check(cmd.CoreSupportsReExec(d), Equals, true)
+}
+
+func (s *cmdSuite) TestInternalToolPathInNewCore(c *C) {
+	d := c.MkDir()
+	defer release.MockOnClassic(true)()
+	defer release.MockReleaseInfo(&release.OS{ID: "ubuntu"})()
+	defer cmd.MockCorePaths(c.MkDir(), d)()
+	defer cmd.MockVersion("2")()
+
+	p := filepath.Join(d, "/usr/lib/snapd/potato")
+	c.Assert(os.MkdirAll(p, 0755), IsNil)
+	c.Assert(ioutil.WriteFile(filepath.Join(d, "/usr/lib/snapd/info"), []byte("VERSION=9999"), 0644), IsNil)
+
+	c.Check(cmd.InternalToolPath("potato"), Equals, p)
+}
+
+func (s *cmdSuite) TestInternalToolPathInOldCore(c *C) {
+	d := c.MkDir()
+	defer release.MockOnClassic(true)()
+	defer release.MockReleaseInfo(&release.OS{ID: "ubuntu"})()
+	defer cmd.MockCorePaths(d, c.MkDir())()
+	defer cmd.MockVersion("2")()
+
+	p := filepath.Join(d, "/usr/lib/snapd/potato")
+	c.Assert(os.MkdirAll(p, 0755), IsNil)
+	c.Assert(ioutil.WriteFile(filepath.Join(d, "/usr/lib/snapd/info"), []byte("VERSION=9999"), 0644), IsNil)
+
+	c.Check(cmd.InternalToolPath("potato"), Equals, p)
+}
+
+func (s *cmdSuite) TestInternalToolPathInBothPrefersNore(c *C) {
+	dNew := c.MkDir()
+	dOld := c.MkDir()
+	defer release.MockOnClassic(true)()
+	defer release.MockReleaseInfo(&release.OS{ID: "ubuntu"})()
+	defer cmd.MockCorePaths(dOld, dNew)()
+	defer cmd.MockVersion("2")()
+
+	pNew := filepath.Join(dNew, "/usr/lib/snapd/potato")
```

Comment:
Given that this is used quite a bit maybe a helper? OTOH, maybe its just not worth it :)